src/app/modules/item/services/item-task-answer.service.ts

```diff
@@ -0,0 +1,100 @@
+        switchMap(([ answer, state ]) => this.updateCurrentAnswerService.update(route.id, attemptId, { answer, state })),
+        switchMap(() => this.getCurrentAnswerService.get(route.id, attemptId)),
+      )),
+    )),
+    shareReplay(1), // avoid duplicate xhr calls on multiple subscriptions.
+  );
+
+  private reloadAnswerAndStateSubject = new Subject<void>();
+  private reloadAnswerAndState$ = combineLatest([ this.currentAnswer$, this.task$ ]).pipe(
+    repeatLatestWhen(this.reloadAnswerAndStateSubject),
+    switchMap(([ currentAnswer, task ]) =>
+      (currentAnswer?.state ? task.reloadState(currentAnswer.state).pipe(mapTo({ currentAnswer, task })) : of({ currentAnswer, task }))
+    ),
+    switchMap(({ currentAnswer, task }) =>
+      (currentAnswer?.answer ? task.reloadAnswer(currentAnswer.answer).pipe(mapTo(undefined)) : of(undefined))
+    ),
+    shareReplay(1),
+  );
+
+  private saveAnswerAndStateInterval$ = this.reloadAnswerAndState$.pipe(
+    switchMap(() => this.task$),
+    switchMap(task => interval(answerAndStateSaveInterval).pipe(mapTo(task))),
+    switchMap(task => forkJoin([ task.getAnswer(), task.getState() ])),
+    distinctUntilChanged(([ answer1, state1 ], [ answer2, state2 ]) => answer1 === answer2 && state1 === state2),
+    skip(1), // avoid saving an answer right after fetching it
+    withLatestFrom(this.config$),
+    switchMap(([ [ answer, state ], { route, attemptId }]) =>
+      this.updateCurrentAnswerService.update(route.id, attemptId, { answer, state })
+    ),
+  );
+
+  private subscriptions = [
+    this.reloadAnswerAndState$.subscribe({ error: err => this.taskInitService.setError(err) }),
+    this.saveAnswerAndStateInterval$.subscribe({ error: err => this.taskInitService.setError(err) }),
+  ];
+
+  constructor(
+    private taskInitService: ItemTaskInitService,
+    private getCurrentAnswerService: GetCurrentAnswerService,
+    private updateCurrentAnswerService: UpdateCurrentAnswerService,
+    private answerTokenService: AnswerTokenService,
+    private gradeService: GradeService,
+  ) {}
+
+  ngOnDestroy(): void {
+    this.subscriptions.forEach(subscription => subscription.unsubscribe());
+  }
+
+  submitAnswer(): Observable<unknown> {
+    return this.task$.pipe(
+      // Step 1: get answer from task
+      switchMap(task => combineLatest([ task.getAnswer(), this.taskToken$ ]).pipe(
+
+        // Step 2: generate answer token with backend
+        switchMap(([ answer, taskToken ]) => this.answerTokenService.generate(answer, taskToken).pipe(
+
+          // Step 3: get answer grade with answer token from task
+          switchMap(answerToken => task.gradeAnswer(answer, answerToken).pipe(
+
+            // Step 4: Save grade in backend
+            switchMap(grade => this.gradeService.save(
+              taskToken,
+              answerToken,
+              grade.score,
+              grade.scoreToken ?? undefined,
+            )),
+          )),
+        )),
+      )),
+    );
```

Comment:
Isn't it a bad smell to create a cascade from a flatmap? What's the reason behind that?

I would a `submitAnswer()` to just submit an answer (one shot, emitting once). Here you return an observable which, in terms of design, might continue saving and so emitting forever.  
Maybe it works as expected, and maybe is safe (I would expect`take(1)`) but anyway, I am not sure I agree with this way of using rx for triggering actions. 🤔 (I think that's not the only place you do that)

Reply:
I agree it is a bad smell, but I need every previous result at each step
And the scenario where I `switchMap(() => xx.pipe(mapTo(…))` is very ugly as well − in my opinion even uglier than cascading flat maps, thus the choice.
As for the `take(1)`, I used it originally, but `task$` is a replay subject that completes after first value

Reply:
<details>
<summary>Proposal 1 − ugly in my opinion</summary>

```ts
return this.task$.pipe(
  take(1),
  // Step 1: get answer from task
  switchMap(task => combineLatest([ task.getAnswer(), this.taskToken$, of(task) ])),

  // Step 2: generate answer token with backend
  switchMap(([ answer, taskToken, task ]) => this.answerTokenService.generate(answer, taskToken).pipe(map(answerToken => ({
    answerToken,
    answer,
    taskToken,
    task,
  })))),

  // Step 3: get answer grade with answer token from task
  switchMap(({ answer, answerToken, taskToken, task }) => task.gradeAnswer(answer, answerToken).pipe(map(grade => ({
    grade,
    taskToken,
    answerToken,
  })))),

  // Step 4: Save grade in backend
  switchMap(({ answerToken, grade, taskToken }) => this.gradeService.save(
    taskToken,
    answerToken,
    grade.score,
    grade.scoreToken ?? undefined,
  ))
);
```

</details>

<details>
<summary>Proposal 2 − OK but less declarative in my opinion</summary>

```ts
const task$ = this.task$.pipe(take(1));

// Step 1: get answer from task
const answer$ = task$.pipe(switchMap(task => task.getAnswer()), shareReplay(1));

// Step 2: generate answer token with backend
const answerToken$ = combineLatest([ this.taskToken$, answer$ ]).pipe(
  switchMap(([ taskToken, answer ]) => this.answerTokenService.generate(answer, taskToken)),
  shareReplay(1),
);

// Step 3: get answer grade with answer token from task
const grade$ = combineLatest([ task$, answer$, answerToken$ ]).pipe(
  switchMap(([ task, answer, answerToken ]) => task.gradeAnswer(answer, answerToken)),
  shareReplay(1),
);

// Step 4: Save grade in backend
return combineLatest([ this.taskToken$, answerToken$, grade$ ]).pipe(
  switchMap(([ taskToken, answerToken, grade ]) => this.gradeService.save(
    taskToken,
    answerToken,
    grade.score,
    grade.scoreToken ?? undefined,
  )),
  shareReplay(1),
);
```

</details>

Reply:
The first one is very readable and can be understood without any effort. 

For the second one, I wouldn't say it is less declarative, you just do declarations and no mutations :-).  
For the readability, the second one, with the variable name as step stones is readable as well. Quite smooth to understand what needs what at the end.

And even conceptually, I like the second proposal as it is more clear that at the end (the return) the outcome of this function is just a `save` (even if there are some processing to get the input of this `save`)

For both cases, the usage of `take(1)` is bit confusing. Here actually, as it is just to do a one-shot-call operation, we want to do it only once (at most), with the value of the observables at the moment of the call. 
If we consider we know when calling get services that they will emit only once, there is just one `take(1)` protection missing for ` this.taskToken$` so that we are sure everything is called at most once.

Are you sure of the last `shareReplay(1)` on the save? It is just for a function call on a button press, that looks not very useful (or it would be missing everywhere in the app).

Reply:
OK, let's go for option 2
No the `shareReplay(1)` is not useful (I realised that afterwards), I'll remove it

Reply:
fixed